File: 2018/nim/day07.nim
import sequtils, sugar, tables

type Workers = object
    MAX: int
    tasks: seq[(char, int)]
    time: int

proc initWorkers(max_workers: int): Workers =
    Workers(MAX: max_workers, time: 0)

proc add(self: var Workers, task: char, task_length: int) =
    if len(self.tasks) == self.MAX:
        return
    for (tt, time) in self.tasks:
        if tt == task:
            return
    self.tasks.add((task, self.time + task_length))

proc ended_tasks(self: var Workers): seq[char] =
    var min_time = high(int)
    for (task, time) in self.tasks:
        min_time = min(time, min_time)
    self.time = min_time
    let rem = collect(newSeq):
        for (task, time) in self.tasks:
            if time == self.time:
                task
    self.tasks = collect(newSeq):
        for (task, time) in self.tasks:
            if task notin rem:
                (task, time)
    return rem

proc remove_task(task_list: var Table[char, seq[char]], task: char) =
    task_list.del(task)
    for k, v in task_list.pairs():
        task_list[k] = v.filter(t => t != task)

var prec = initTable[char, seq[char]]()
for line in lines("data/input-07.txt"):
    let (bef, aft) = (line[5], line[36])
    if not prec.hasKey(bef):
        prec[bef] = newSeq[char]()
    if not prec.hasKey(aft):
        prec[aft] = newSeq[char]()
    prec[aft].add(bef)

var (prec1, res1) = (prec, "")
while len(prec1) > 0:
    var all_tasks = collect(newSeq):
        for k, v in prec1.mpairs():
            (len(v), k)
    let task = min(all_tasks)[1]
    prec1.remove_task(task)
    res1.add(task)
echo "Part 1: ", res1
assert(res1 == "BCADPVTJFZNRWXHEKSQLUYGMIO")

var (workers, res2) = (initWorkers(5), 0)
while len(prec) > 0:
    let next_tasks = collect(newSeq):
        for (k, v) in prec.pairs():
            if len(v) == 0:
                k
    for task in next_tasks:
        workers.add(task, 60 + ord(task) - ord('A') + 1)
    for task in workers.ended_tasks():
        prec.remove_task(task)
        res2 = workers.time
echo "Part 2: ", res2
assert(res2 == 973)

File: 2018/src/day07.rs
use std::collections::HashMap;
use std::fs;

struct Task {
    name: char,
    endtime: usize,
}
struct Workers {
    tasks: Vec<Task>,
    time: usize,
}
impl Workers {
    const MAX: usize = 5;
    fn new() -> Self {
        Workers {
            tasks: Vec::new(),
            time: 0,
        }
    }
    fn add(&mut self, task: char, length: usize) {
        if (self.tasks.len() == Workers::MAX) || self.tasks.iter().any(|z| z.name == task) {
            return;
        }
        self.tasks.push(Task {
            name: task,
            endtime: self.time + length,
        });
    }
    fn ended_tasks(&mut self) -> Vec<char> {
        self.tasks.sort_unstable_by_key(|z| z.endtime);
        self.time = self.tasks[0].endtime;
        match self.tasks.iter().rposition(|z| z.endtime == self.time) {
            Some(pos) => self.tasks.drain(0..=pos).map(|z| z.name).collect(),
            None => Vec::new(),
        }
    }
}
fn remove_task(list: &mut HashMap<char, Vec<Option<char>>>, task: char) {
    list.remove(&task);
    list.iter_mut()
        .for_each(|z| z.1.retain(|&x| x != Some(task)));
}

pub fn day07() {
    let input = "data/input-07.txt";
    let file = fs::read_to_string(input)
        .unwrap()
        .replace("Step ", "")
        .replace(" must be finished before step ", "")
        .replace(" can begin.", "");
    let mut prec = HashMap::new();
    for line in file.lines() {
        if let [bef, aft] = line.chars().collect::<Vec<_>>()[..] {
            prec.entry(bef).or_insert(Vec::new());
            prec.entry(aft).or_insert(Vec::new()).push(Some(bef));
        }
    }

    let (mut prec1, mut res) = (prec.clone(), String::new());
    while let Some((&task, _)) = prec1
        .iter()
        .min_by_key(|&(&k, ref v)| v.len() * 1000 + k as usize)
    {
        remove_task(&mut prec1, task);
        res.push(task);
    }
    println!("Part 1: {res}");
    assert_eq!(res, "BCADPVTJFZNRWXHEKSQLUYGMIO");

    let mut workers = Workers::new();
    while !prec.is_empty() {
        let next_tasks = prec.iter().filter(|(_, v)| v.is_empty()).map(|(k, _)| k);
        for &task in next_tasks {
            workers.add(task, 60 + task as usize - 'A' as usize + 1);
        }
        for task in workers.ended_tasks() {
            remove_task(&mut prec, task);
        }
    }
    let res = workers.time;
    println!("Part 2: {res}");
    assert_eq!(res, 973);
}
